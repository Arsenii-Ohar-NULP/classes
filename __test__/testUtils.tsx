import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
