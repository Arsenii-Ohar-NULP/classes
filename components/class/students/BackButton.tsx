import React from 'react';
import {useRouter} from "next/router";
import {Button} from "react-bootstrap";
import backIcon from 'icons/back.svg';
import Image from "next/image";

export const BackButton = () => {
    const router = useRouter();

    const goBack = async () => {
        router.back();
    }

    return <Button variant={'outline-light'} className={'p-1 rounded-pill'} onClick={goBack}>
        <Image src={backIcon} alt={'Back'} height={24} width={24} />
    </Button>
}