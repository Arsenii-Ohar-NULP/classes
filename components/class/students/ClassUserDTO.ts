export interface ClassUserDTO{
        class: number;
        student: number;
}

