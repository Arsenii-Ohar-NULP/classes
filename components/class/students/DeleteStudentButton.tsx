import React from 'react';
import {Button} from "react-bootstrap";
import deleteIcon from 'icons/delete.svg';
import Image from "next/image";

interface DeleteStudentButtonProps {
    onDelete: VoidFunction
}

export const DeleteStudentButton = ({onDelete}: DeleteStudentButtonProps) => {
    return <Button onClick={onDelete} variant={'danger'} className={'p-1'}>
        <Image src={deleteIcon} alt={'Delete Icon'} height={24} width={24}/>
    </Button>
}