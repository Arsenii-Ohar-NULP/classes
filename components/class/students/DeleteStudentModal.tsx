import React, {useState} from "react";
import {useRouter} from "next/navigation";
import {useAppDispatch} from "components/redux/store";
import InvalidCredentials from "components/errors/InvalidCredentials";
import {logout} from "components/login/AuthService";
import {Button, Modal} from "react-bootstrap";
import {ClassUserDTO, removeStudentFromClass} from "components/class/students/StudentsService";

export default function DeleteStudentModal({
  classUser,
  onDelete,
  show,
  close,
}: {
  classUser: ClassUserDTO;
  onDelete: () => void;
  show: boolean;
  close: () => void;
}) {
  const [isDeleting, setIsDeleting] = useState<boolean>(false);

  const router = useRouter();
  const dispatch = useAppDispatch();

  function deleteStudent() {
    setIsDeleting(true);
    removeStudentFromClass(classUser)
      .then(() => {
        onDelete();
        close();
      })
      .catch((error) => {
        if (error instanceof InvalidCredentials) {
          logout(dispatch, router);
        }
      })
      .finally(() => setIsDeleting(false));
  }

  return (
    <>
      <Modal show={show} backdrop={true} onHide={close} keyboard={false}>
        <Modal.Header closeButton={!isDeleting}>
          <Modal.Title>Are you sure?</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          The student will be deleted and there is no way to undo this. Are you
          sure?
        </Modal.Body>
        <Modal.Footer>
          <Button
            variant="danger"
            onClick={deleteStudent}
            disabled={isDeleting}
          >
            Delete
          </Button>
          <Button variant="success" onClick={close} disabled={isDeleting}>
            No, take me back
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
