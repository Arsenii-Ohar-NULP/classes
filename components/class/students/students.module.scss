/* Hide scrollbar for Chrome, Safari and Opera */
.hide-scrollbar::-webkit-scrollbar {
  display: none;
}

/* Hide scrollbar for IE, Edge and Firefox */
.hide-scrollbar {
  -ms-overflow-style: none;  /* IE and Edge */
  scrollbar-width: none;  /* Firefox */
}

.smooth-shadow{
  box-shadow: rgba(47, 43, 61, 0.1) 0px 4px 18px 0px;
}

.neat-text {
  color: rgba(47, 43, 61, 0.78);
}

.student{}

.student:hover{
  background-color: #eafdd8;
}