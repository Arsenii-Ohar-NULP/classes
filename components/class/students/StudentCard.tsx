import React, {useState} from 'react';
import User from "../../account/User";
import clsx from "clsx";
import ProfilePicture from "../../ProfilePicture";
import styles from 'components/class/students/students.module.scss';
import {DeleteStudentButton} from "./DeleteStudentButton";
import {VisibleOnHover} from "./VisibleOnHover";

interface StudentCardProps {
    student: User;
    onDelete: VoidFunction
}

export const StudentCard = ({student, onDelete}: StudentCardProps) => {
    const [isHover, setIsHover] = useState<boolean>();
    return <tr
        className={clsx('p-2 w-100 text-start overflow-scroll', styles['neat-text'], styles['student'])}
        onMouseOver={() => setIsHover(true)}
        onMouseOut={() => setIsHover(false)}>
        <td scope={'row'} className={'d-flex text-center gap-2 align-items-center'}>
            <ProfilePicture user={student} hoverOn={false}/>
            <div>{student.firstName} {student.lastName}</div>
        </td>
        <td scope={'row'}>{student.email}</td>
        <td scope={'row'}>
            <div className={'d-flex gap-3'}>
                {student.phone}
                <VisibleOnHover isHover={isHover}>
                    <DeleteStudentButton onDelete={onDelete}/>
                </VisibleOnHover>
            </div>
        </td>
    </tr>
}