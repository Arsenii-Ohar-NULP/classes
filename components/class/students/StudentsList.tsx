import React, {useState} from 'react';
import User from "components/account/User";
import {StudentCard} from "components/class/students/StudentCard";
import {StudentsListHeader} from "components/class/students/StudentsListHeader";
import clsx from "clsx";
import styles from 'components/class/students/students.module.scss';
import {useAppSelector} from "../../redux/store";
import Class from "components/classes/Class";
import DeleteStudentModal from "./DeleteStudentModal";

interface StudentsListProps {
    students: User[];
    isLoading: boolean;
    cls: number;
    onDelete: (student: number) => void;
}

export const StudentsList = ({students, isLoading, cls, onDelete}: StudentsListProps) => {
    const [deleteStudentId, setDeleteStudentId] = useState<number>(-1);

    return (<div className={'container'}>
        {!isLoading && (students == null || students.length == 0) ?
            <p className={'text-center'}>There are no students in this class.</p> : !isLoading &&
            <div className={clsx('overflow-scroll', styles['hide-scrollbar'], styles['smooth-shadow'])}>
                <table className={'table shadow m-0'}>
                    <StudentsListHeader/>
                    <tbody>
                    {students?.map((student) => <StudentCard key={student.id}  student={student} onDelete={() => setDeleteStudentId(student.id)}/>)}
                    </tbody>
                </table>
                <DeleteStudentModal
                    classUser={{student: deleteStudentId, class: cls}}
                    show={deleteStudentId != -1}
                    onDelete={() => onDelete(deleteStudentId)}
                    close={() => setDeleteStudentId(-1)}/>
            </div>
        }
    </div>)
}