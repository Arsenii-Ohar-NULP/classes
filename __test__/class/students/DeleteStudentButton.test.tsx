import React from 'react';
import {renderWithProviders} from "../../testUtils";
import {DeleteStudentButton} from "../../../components/class/students/DeleteStudentButton";
import {fireEvent, screen, waitFor} from "@testing-library/react";

describe('Delete Student Button', () => {
    it('should match a snapshot', () => {
        const button = renderWithProviders(<DeleteStudentButton onDelete={jest.fn()}/>);
        expect(button).toMatchSnapshot();
    })
    it('should call onDelete, when clicked', async () => {
        const onDelete = jest.fn();
        renderWithProviders(<DeleteStudentButton onDelete={onDelete}/>);
        const deleteButton = screen.getByRole('button');
        fireEvent.click(deleteButton);

        await waitFor(() => {
            expect(onDelete).toHaveBeenCalled();
        })
    })
})