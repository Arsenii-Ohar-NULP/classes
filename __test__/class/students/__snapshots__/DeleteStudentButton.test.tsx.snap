// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete Student Button should match a snapshot 1`] = `
<div>
  <button
    class="p-1 btn btn-danger"
    type="button"
  >
    <img
      alt="Delete Icon"
      data-nimg="1"
      decoding="async"
      height="24"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
      style="color: transparent;"
      width="24"
    />
  </button>
</div>
`;
